pkg/pillar/cmd/zedrouter/appnumonunet.go

```diff
@@ -0,0 +1,303 @@
+			appNum = int(addr - start)
+		} else {
+			// TBD:XXX, this loop needs to be optimized
+			i := 0
+			for i < netInstStatus.DhcpAppNumCount {
+				if netInstStatus.AppNumBitMap.IsSet(i) {
+					i++
+					continue
+				}
+				appNum = i
+				log.Functionf("uNetAppNumAlloc(%s-%s): app(%s) got %d\n",
+					netInstStatus.DisplayName, netInstStatus.UUID, appID,
+					appNum)
+				break
+			}
+			if i > netInstStatus.DhcpAppNumCount {
+				errStr := fmt.Sprintf("no free slot")
+				log.Errorf("uNetAppNumAlloc(%s-%s): app(%s) fail: %s\n",
+					netInstStatus.DisplayName, netInstStatus.UUID.String(),
+					appID, errStr)
+				return types.AppNumInvalid, errors.New(errStr)
+			}
+		}
+	}
+
+	// check whether the App Num is already assigned to someone else
+	if networkInstanceIsDuplicateIP(netInstStatus, appNum, mac) {
+		errStr := fmt.Sprintf("duplicate IP")
+		log.Errorf("uNetAppNumAlloc(%s-%s): app(%s) fail: %s\n",
+			netInstStatus.DisplayName, netInstStatus.UUID.String(),
+			appID, errStr)
+		return types.AppNumInvalid, errors.New(errStr)
+	}
+	return appNum, nil
+}
+
+func uNetAppNumDelete(ctx *zedrouterContext,
+	netInstStatus *types.NetworkInstanceStatus, appID string, appNum int) {
+	bitMapAppNumDelete(ctx, netInstStatus, appNum)
+}
+
+// App Num BitMap Routines
+func bitMapAppNumAdd(ctx *zedrouterContext,
+	netInstStatus *types.NetworkInstanceStatus, appNum int) {
+	if appNum >= types.AppNumStart && netInstStatus != nil &&
+		appNum <= netInstStatus.DhcpAppNumCount &&
+		!netInstStatus.AppNumBitMap.IsSet(appNum) {
+		netInstStatus.AppNumBitMap.Set(appNum)
+	}
+}
+
+func bitMapAppNumDelete(ctx *zedrouterContext,
+	netInstStatus *types.NetworkInstanceStatus, appNum int) {
+	if appNum >= types.AppNumStart && netInstStatus != nil &&
+		appNum <= netInstStatus.DhcpAppNumCount &&
+		netInstStatus.AppNumBitMap.IsSet(appNum) {
+		netInstStatus.AppNumBitMap.Clear(appNum)
+	}
+}
+
+// Persistent Address Map routines
+func persistentAddrMapGet(ctx *zedrouterContext, networkID string,
+	appID string) *types.NetworkInstanceAppNumMap {
+	pub := ctx.pubNetworkInstanceAppNumMap
+	key := persistentAddrMapKey(networkID, appID)
+	if addrMap, _ := pub.Get(key); addrMap != nil {
+		appNumMap := addrMap.(types.NetworkInstanceAppNumMap)
+		return &appNumMap
+	}
+	return nil
+}
+
+func persistentAddrMapAdd(ctx *zedrouterContext,
+	networkID string, appID string, appNumOnUNet int) {
+	pub := ctx.pubNetworkInstanceAppNumMap
+	addrMap := types.NetworkInstanceAppNumMap{
+		NetworkID:    networkID,
+		AppID:        appID,
+		AppNumOnUNet: appNumOnUNet,
+	}
+	pub.Publish(addrMap.Key(), addrMap)
+	return
+}
+
+func persistentAddrMapDelete(ctx *zedrouterContext,
+	networkID string, appID string) {
+	pub := ctx.pubNetworkInstanceAppNumMap
+	key := persistentAddrMapKey(networkID, appID)
+	pub.Unpublish(key)
+	return
+}
+
+// Persistent IPAddress Mapping Restore for Network Instance
+func persistentAddrMapRestoreForNetworkInstance(ctx *zedrouterContext,
```

Comment:
It would be much easier if this wasn't unique code but either followed the identical pattern (and I really mean copy and paste + change key type + rename functions and no other chages) from appnum/bridgenum allocator code.